#50 don't exchange with recently-seen peers

GitMeGet opened the pull request.

GitMeGet committed:
WIP: store last seen peers

GitMeGet committed:
continue WIP

- timestamp when peer was last seen
- update hashmap tempid after reading from peer
- TODO:
  -- wrap functions as class
  -- remove hashmap entries if > 15min ago
  -- call BLEAddress::toString() less?

GitMeGet committed:
refactor into PeerCache class

GitMeGet committed:
compare time using uint32_t instead of uint8_t

- use year,month,day,hour,minute from TS_DateTime

GitMeGet committed:
change _PeerCache from member class to singleton

GitMeGet committed:
task to cleanup outdated hashmap values

GitMeGet committed:
cleanup code

jamesi8086 commented on alpha/hal.cpp:
```diff
@@ -272,6 +272,17 @@ void _TS_HAL::rtc_get(TS_DateTime &dt)
   EXIT_CRITICAL;
 }
 
+uint32_t _TS_HAL::rtc_get_mins(void)
```
this should go into cleanbox if it doesn't already exist

GitMeGet replied:
into cleanbox.h?

jamesi8086 replied:
yeah

jamesi8086 commented on alpha/hal.h:
```diff
@@ -122,6 +122,7 @@ class _TS_HAL
     // RTC
     //
     void rtc_get(TS_DateTime &);
+    uint32_t rtc_get_mins(void);
```
as above

yongkimleng commented on alpha/opentracev2.h:
```diff
@@ -53,6 +59,22 @@ struct OT_ConnectionRecord
 // Classes
 //
 
+class _PeerCache
```
not a singleton, can be PeerCache

jamesi8086 replied:
if class is specific to _OT_ProtocolV2, it could be _OT_ProtocolV2_PeerCache

jamesi8086 commented on alpha/opentracev2.h:
```diff
@@ -53,6 +59,22 @@ struct OT_ConnectionRecord
+class _PeerCache
+{
+  public:
+    std::string shouldConnect(std::string BLEAddrStr, uint32_t currMin);
```
use by ref (&) instead of by val

jamesi8086 replied:
std::string & is faster when no copying of strings is required

jamesi8086 replied:
lowercase argument names

jamesi8086 commented on alpha/opentracev2.cpp:
```diff
@@ -54,6 +118,25 @@ void _OT_ProtocolV2::begin()
   this->bleAdvertising->setMinPreferred(0x06);  // functions that help with iPhone connections issue
   // TODO: find out what setMin and setMax interval really means
 
+  xTaskCreatePinnedToCore(
```
no need to create new task to cleanup, we could track timestamp of last cleanup and run it sequentially during operational calls topeercache

jamesi8086 replied:
in fact tasks could cause concurrency issues

GitMeGet replied:
run it at the end of _OT_ProtocolV2::scan_and_connect()?

jamesi8086 replied:
yes, or at the start. 
see how its implemented in storage.h, the cleanup function does a bunch of checks.

jamesi8086 reviewed (approved):
actually this looks really similar to what was done in Storage.

wonder if Storage should actually add BLE address there.

GitMeGet committed:
Merge branch 'master' into yp-#50

GitMeGet committed:
Merge remote-tracking branch 'upstream/master' into yp-#50

GitMeGet committed:
address code review comments

- pass objects by reference instead of value
- change class name to be more specific
- change argument names to lowercase
